### ISSUE TITLE
Bug: [switch-exhaustiveness-check] False positive with noUncheckedIndexedAccess

### ISSUE THREAD

**Issue by jtbandes** [typescript-eslint/typescript-eslint#10228]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.5&fileType=.tsx&code=GYVwdgxgLglg9mABMOcAUAPAXIgzlAJxjAHMBtAXQEpEBvAKESbwHcYoIALRTMgBmp1GzJhACGuAKaIARJxgyswkUwBGBSWIDWAbmXNxUxOAAmk4MUkmlKxGeBiQAGyg3biKJwJwWiMJN8AUQJvAjQAAwBVfwwAB0loK0QANzEnEEkcABJaDH4KAF9wqj0RAvoCoA&eslintrc=N4KABGBEAOCGBOBnApvSAuKABALgT2mUQGN4BLaHAWiIBsyA7HAejiVUgBpwp4BXWkQxhQECJFwEipCtTqMWiAO5kcxABY0AHuth9EOMgDdkDIoiobkxANbDIqeAHs0PAL4g3QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2RGwAqkWgALdNADW6ACYBJIjPQAPWQEFo0dCTKUO6XgF8QhoA&tokens=false

### Repro Code

```TypeScript
function foo(x: string[]) {
    switch (x[0]) {
      case "hi":
        break;
      case undefined:
      default:
        throw new Error(`Unexpected value: ${x[0]}`);
    }
}
```

### ESLint Config

```javascript
module.exports = {
  "parser": "@typescript-eslint/parser",
  "rules": {
    "@typescript-eslint/switch-exhaustiveness-check": "error"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true,
    "noUncheckedIndexedAccess": true
  }
}
```

### Expected Result

No error

### Actual Result

"Switch is not exhaustive. Cases not matched: undefined" — even though `case undefined:` is present

**Comment by bradzacher** [comment]
The new rule options had an incorrect default and has been fixed in #10223 and has been released in 8.12.2

**Comment by jtbandes** [comment]
I experienced this bug with 8.12.2 (also repro'd with 8.12.2 in the playground)

**Comment by bradzacher** [comment]
Ah wait my playground hadn't updated to 8.12 sorry. It repros now that I've cleared the cache.

**Comment by notaphplover** [comment]
Hey @bradzacher, I'm having issues with this rule when facing switch code flows with a `default` case when switching from an enum typed variable:

```ts
enum NumberEnum {
  one = 1,
  zero = 0,
}

function getNumber(): NumberEnum {
  return 0;
}

function isZeroGenerated(): boolean {
  const foo: NumberEnum = getNumber();

  // Switch is not exhaustive. Cases not matched: NumberEnum.one
  switch (foo) {
    case NumberEnum.zero:
      return true;
    default:
      return false;
  }
}
```

Should I open a new issue?

**Comment by kirkwaiblinger** [comment]
@notaphplover This is an intentional change, see https://github.com/typescript-eslint/typescript-eslint/issues/8247, https://github.com/typescript-eslint/typescript-eslint/pull/9954, https://github.com/typescript-eslint/typescript-eslint/issues/10222. If you'd like to opt in to the old behavior, you can use the option `considerDefaultExhaustiveForUnions: true`. See also [the updated rule doc page](https://typescript-eslint.io/rules/switch-exhaustiveness-check/#considerdefaultexhaustiveforunions)
